Interrupts break class loading

If a thread is [interrupted](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/Thread.html#interrupt()) shortly before or while it is loading a class from disk, any further class loads from that jar will fail.

To reproduce, run this snippet any time before JoinMultiplayerScreen (arbitrary, could be any class) is naturally loaded by the game, then do anything that requires it or any other class from the same jar file to be loaded (in case of the minecraft jar, that's pretty much anything that you haven't done before).
```java
        new Thread(() -> {
            Thread.currentThread().interrupt();
            JoinMultiplayerScreen.class.getName();
        }).start();
```
<details>
<summary>Crash</summary>

```
Exception in thread "Thread-5" 
java.lang.NoClassDefFoundError: net/minecraft/client/gui/screens/multiplayer/JoinMultiplayerScreen
	at TRANSFORMER/test@1.0.0/com.example.ExampleMod.lambda$init$2(ExampleMod.java:190)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ClassNotFoundException: net.minecraft.client.gui.screens.multiplayer.JoinMultiplayerScreen
	at cpw.mods.securejarhandler/cpw.mods.cl.ModuleClassLoader.loadClass(ModuleClassLoader.java:141)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 2 more
```
followed shortly by a fatal
```
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.polymc.impl.OneSixLauncher.invokeMain(OneSixLauncher.java:104)
	at org.polymc.impl.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:175)
	at org.polymc.impl.OneSixLauncher.launch(OneSixLauncher.java:185)
	at org.polymc.EntryPoint.listen(EntryPoint.java:144)
	at org.polymc.EntryPoint.main(EntryPoint.java:74)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.github.zekerzhayard.forgewrapper.installer.Main.main(Main.java:57)
	... 9 more
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:32)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.Launcher.run(Launcher.java:106)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.Launcher.main(Launcher.java:77)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23)
	at cpw.mods.bootstraplauncher@1.1.2/cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:141)
	... 14 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at MC-BOOTSTRAP/fmlloader@1.19.2-43.1.34/net.minecraftforge.fml.loading.targets.CommonClientLaunchHandler.lambda$launchService$0(CommonClientLaunchHandler.java:27)
	at MC-BOOTSTRAP/cpw.mods.modlauncher@10.0.8/cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:30)
	... 21 more
Caused by: java.lang.NoClassDefFoundError: net/minecraft/CrashReportCategory
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.CrashReport.m_127516_(CrashReport.java:157)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.CrashReport.m_127514_(CrashReport.java:153)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.client.main.Main.m_239872_(Main.java:183)
	at TRANSFORMER/minecraft@1.19.2/net.minecraft.client.main.Main.main(Main.java:51)
	... 27 more
Caused by: java.lang.ClassNotFoundException: net.minecraft.CrashReportCategory
	at cpw.mods.securejarhandler/cpw.mods.cl.ModuleClassLoader.loadClass(ModuleClassLoader.java:141)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 31 more
```

</details>

To understand why exactly this happens, it's helpful to add some logging to this try-catch-ignore:
https://github.com/McModLauncher/securejarhandler/blob/ee35db45b3da48a7ff1b844313167d0a8e5019b0/src/main/java/cpw/mods/cl/ModuleClassLoader.java#L218-L222
which will then reveal this much more helpful stacktrace:
<details>
<summary>Stacktrace</summary>

```
java.nio.channels.ClosedByInterruptException
	at java.base/java.nio.channels.spi.AbstractInterruptibleChannel.end(AbstractInterruptibleChannel.java:199)
	at java.base/sun.nio.ch.FileChannelImpl.endBlocking(FileChannelImpl.java:168)
	at java.base/sun.nio.ch.FileChannelImpl.readInternal(FileChannelImpl.java:820)
	at java.base/sun.nio.ch.FileChannelImpl.read(FileChannelImpl.java:800)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.readFullyAt(ZipFileSystem.java:1236)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.readFullyAt(ZipFileSystem.java:1231)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem$EntryInputStream.initDataPos(ZipFileSystem.java:2263)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem$EntryInputStream.read(ZipFileSystem.java:2205)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem$2.fill(ZipFileSystem.java:2155)
	at java.base/java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
	at java.base/java.io.InputStream.readNBytes(InputStream.java:409)
	at java.base/java.io.InputStream.readAllBytes(InputStream.java:346)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.newByteChannel(ZipFileSystem.java:972)
	at jdk.zipfs/jdk.nio.zipfs.ZipPath.newByteChannel(ZipPath.java:864)
	at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newByteChannel(ZipFileSystemProvider.java:238)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:380)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:432)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.niofs.union.UnionFileSystem.byteChannel(UnionFileSystem.java:228)
	at java.base/java.util.Optional.map(Optional.java:260)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.niofs.union.UnionFileSystem.newReadByteChannel(UnionFileSystem.java:219)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.niofs.union.UnionFileSystemProvider.newByteChannel(UnionFileSystemProvider.java:149)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:380)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:432)
	at java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:422)
	at java.base/java.nio.file.Files.newInputStream(Files.java:160)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.JarModuleFinder$JarModuleReader.lambda$open$0(JarModuleFinder.java:78)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.util.LambdaExceptionUtils.lambda$rethrowFunction$2(LambdaExceptionUtils.java:42)
	at java.base/java.util.Optional.map(Optional.java:260)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.JarModuleFinder$JarModuleReader.open(JarModuleFinder.java:78)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.lambda$getClassBytes$11(ModuleClassLoader.java:106)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.util.LambdaExceptionUtils.lambda$rethrowFunction$2(LambdaExceptionUtils.java:42)
	at java.base/java.util.Optional.flatMap(Optional.java:289)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.getClassBytes(ModuleClassLoader.java:106)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.readerToClass(ModuleClassLoader.java:114)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.lambda$findClass$16(ModuleClassLoader.java:236)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.loadFromModule(ModuleClassLoader.java:247)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.findClass(ModuleClassLoader.java:236)
	at cpw.mods.securejarhandler@1.0.3/cpw.mods.cl.ModuleClassLoader.loadClass(ModuleClassLoader.java:152)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at TRANSFORMER/test@1.0.0/com.example.ExampleMod.lambda$init$2(ExampleMod.java:190)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

</details>

And as per the [documentation of `ClosedByInterruptException`](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/ClosedByInterruptException.html), "Before this exception is thrown the channel will have been closed". Meaning the ZipFileSystem is effectively closed at that point and any further reads will result in a `ClosedChannelException` which is why all the other classes then also fail to load.

The issue can also be triggered with just Forge on its own given some timing luck, e.g. the lan server broadcast listener thread that's active in the JoinMultiplayerScreen gets interrupted by the game when you leave that screen, and if that interrupt happens after the first broadcast has been received but before it's processed, then the processing will involve class loads, triggering the same issue (usually breaking the guava).

I've been testing with securejarhandler 2.1.4 (Forge 43.1.34 / Minecraft 1.19.2) and securejarhandler 1.0.3 (Forge 40.1.80 / Minecraft 1.18.2) but this presumably affects all ModLauncher versions. Using [OpenJDK 17.0.1 from flathub](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk17) (via PolyMC).

--- a/src/main/java/cpw/mods/cl/UnionURLStreamHandler.java
+++ b/src/main/java/cpw/mods/cl/UnionURLStreamHandler.java
@@ -1,5 +1,7 @@
 package cpw.mods.cl;
 
+import cpw.mods.niofs.union.UnionPath;
+
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.UncheckedIOException;
@@ -19,11 +21,13 @@ public class UnionURLStreamHandler implements ModularURLHandler.IURLProvider {
     public Function<URL, InputStream> inputStreamFunction() {
         return u-> {
             try {
-                return Files.newInputStream(Paths.get(u.toURI()));
+                if (Paths.get(u.toURI()) instanceof UnionPath upath) {
+                    return upath.buildInputStream();
+                } else {
+                    throw new IllegalArgumentException("Invalid Path "+u.toURI()+" at UnionURLStreamHandler");
+                }
             } catch (URISyntaxException e) {
                 throw new RuntimeException(e);
-            } catch (IOException e) {
-                throw new UncheckedIOException(e);
             }
 
 

--- a/src/main/java/cpw/mods/niofs/union/UnionFileSystem.java
+++ b/src/main/java/cpw/mods/niofs/union/UnionFileSystem.java
@@ -6,6 +6,7 @@ import java.lang.invoke.MethodHandles;
 import java.lang.invoke.MethodType;
 import java.lang.reflect.Method;
 import java.nio.channels.ClosedByInterruptException;
+import java.nio.channels.FileChannel;
 import java.nio.channels.InterruptibleChannel;
 import java.nio.channels.SeekableByteChannel;
 import java.nio.file.AccessMode;
@@ -114,7 +115,7 @@ public class UnionFileSystem extends FileSystem {
         return this.key;
     }
 
-    private record EmbeddedFileSystemMetadata(Path path, FileSystem fs) {}
+    private record EmbeddedFileSystemMetadata(Path path, FileSystem fs, SeekableByteChannel fsCh) {}
 
     public UnionFileSystem(final UnionFileSystemProvider provider, final BiPredicate<String, String> pathFilter, final String key, final Path... basepaths) {
         this.pathFilter = pathFilter;
@@ -133,9 +134,9 @@ public class UnionFileSystem extends FileSystem {
     private static Optional<EmbeddedFileSystemMetadata> openFileSystem(final Path path) {
         try {
             var zfs = FileSystems.newFileSystem(path);
-            Object fci = ZIPFS_CH.invoke(zfs);
+            FileChannel fci = (FileChannel) ZIPFS_CH.invoke(zfs);
             FCI_UNINTERUPTIBLE.invoke(fci);
-            return Optional.of(new EmbeddedFileSystemMetadata(path, zfs));
+            return Optional.of(new EmbeddedFileSystemMetadata(path, zfs, fci));
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         } catch (Throwable t) {
@@ -225,7 +226,7 @@ public class UnionFileSystem extends FileSystem {
         try {
             if (path.getFileSystem() == FileSystems.getDefault() && !path.toFile().exists()) {
                 return Optional.empty();
-            } else if (path.getFileSystem().provider().getScheme().equals("jar") && !zipFsExists(path)) {
+            } else if (path.getFileSystem().provider().getScheme().equals("jar") && !zipFsExists(this, path)) {
                 return Optional.empty();
             } else {
                 return Optional.of(path.getFileSystem().provider().readAttributes(path, BasicFileAttributes.class));
@@ -243,8 +244,9 @@ public class UnionFileSystem extends FileSystem {
                 .findFirst();
     }
 
-    private static boolean zipFsExists(Path path) {
+    private static boolean zipFsExists(UnionFileSystem ufs, Path path) {
         try {
+            if (Optional.ofNullable(ufs.embeddedFileSystems.get(path.getFileSystem())).filter(efs->!efs.fsCh.isOpen()).isPresent()) throw new IllegalStateException("The zip file has closed!");
             return (boolean) ZIPFS_EXISTS.invoke(path);
         } catch (Throwable t) {
             throw new IllegalStateException(t);
@@ -259,7 +261,7 @@ public class UnionFileSystem extends FileSystem {
                         return Optional.of(realPath);
                     }
                 } else if (realPath.getFileSystem().provider().getScheme().equals("jar")) {
-                    if (zipFsExists(realPath)) {
+                    if (zipFsExists(this, realPath)) {
                         return Optional.of(realPath);
                     }
                 } else if (Files.exists(realPath)) {
@@ -355,7 +357,7 @@ public class UnionFileSystem extends FileSystem {
                 continue;
             } else if (dir.getFileSystem() == FileSystems.getDefault() && !dir.toFile().exists()) {
                 continue;
-            } else if (dir.getFileSystem().provider().getScheme() == "jar" && !zipFsExists(dir)) {
+            } else if (dir.getFileSystem().provider().getScheme() == "jar" && !zipFsExists(this, dir)) {
                 continue;
             } else if (Files.notExists(dir)) {
                 continue;

--- a/src/main/java/cpw/mods/niofs/union/UnionFileSystem.java
+++ b/src/main/java/cpw/mods/niofs/union/UnionFileSystem.java
@@ -1,10 +1,12 @@
 package cpw.mods.niofs.union;
 
-import java.io.FileNotFoundException;
-import java.io.IOException;
+import java.io.*;
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodHandles;
 import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
+import java.nio.channels.ClosedByInterruptException;
+import java.nio.channels.InterruptibleChannel;
 import java.nio.channels.SeekableByteChannel;
 import java.nio.file.AccessMode;
 import java.nio.file.DirectoryStream;
@@ -37,7 +39,11 @@ import java.util.stream.StreamSupport;
 
 public class UnionFileSystem extends FileSystem {
     private static final MethodHandle ZIPFS_EXISTS;
+    private static final MethodHandle ZIPFS_CH;
+    private static final MethodHandle FCI_UNINTERUPTIBLE;
     static final String SEP_STRING = "/";
+
+
 
     static {
         try {
@@ -47,10 +53,27 @@ public class UnionFileSystem extends FileSystem {
 
             var clz = Class.forName("jdk.nio.zipfs.ZipPath");
             ZIPFS_EXISTS = hack.findSpecial(clz, "exists", MethodType.methodType(boolean.class), clz);
+
+            clz = Class.forName("jdk.nio.zipfs.ZipFileSystem");
+            ZIPFS_CH = hack.findGetter(clz, "ch", SeekableByteChannel.class);
+
+            clz = Class.forName("sun.nio.ch.FileChannelImpl");
+            FCI_UNINTERUPTIBLE = hack.findSpecial(clz, "setUninterruptible", MethodType.methodType(void.class), clz);
         } catch (NoSuchFieldException | IllegalAccessException | ClassNotFoundException | NoSuchMethodException e) {
             throw new RuntimeException(e);
         }
     }
+
+    public InputStream buildInputStream(final UnionPath path) {
+        try {
+            var bytes = Files.readAllBytes(path);
+            return new ByteArrayInputStream(bytes);
+        } catch (IOException ioe)
+        {
+            throw new UncheckedIOException(ioe);
+        }
+    }
+
     private static class NoSuchFileException extends java.nio.file.NoSuchFileException {
         public NoSuchFileException(final String file) {
             super(file);
@@ -109,9 +132,14 @@ public class UnionFileSystem extends FileSystem {
 
     private static Optional<EmbeddedFileSystemMetadata> openFileSystem(final Path path) {
         try {
-            return Optional.of(new EmbeddedFileSystemMetadata(path, FileSystems.newFileSystem(path)));
+            var zfs = FileSystems.newFileSystem(path);
+            Object fci = ZIPFS_CH.invoke(zfs);
+            FCI_UNINTERUPTIBLE.invoke(fci);
+            return Optional.of(new EmbeddedFileSystemMetadata(path, zfs));
         } catch (IOException e) {
             throw new UncheckedIOException(e);
+        } catch (Throwable t) {
+            throw new IllegalStateException(t);
         }
     }
 

--- a/src/main/java/cpw/mods/niofs/union/UnionPath.java
+++ b/src/main/java/cpw/mods/niofs/union/UnionPath.java
@@ -1,6 +1,7 @@
 package cpw.mods.niofs.union;
 
 import java.io.IOException;
+import java.io.InputStream;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.nio.file.LinkOption;
@@ -324,5 +325,9 @@ public class UnionPath implements Path {
     @Override
     public String toString() {
         return (this.absolute ? UnionFileSystem.SEP_STRING : "") + String.join(UnionFileSystem.SEP_STRING, this.pathParts);
+    }
+
+    public InputStream buildInputStream() {
+        return fileSystem.buildInputStream(this);
     }
 }

